src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java

```diff
@@ -57,18 +59,33 @@ public void visitCreatePath(CreatePath cp) {
             String parentPath = parentPathBuilder.toString();
             final String fullPath = String.format("%s/%s", parentPath, psd.getSegment());
             try {
-                if (session.itemExists(fullPath)) {
-                    log.info("Path already exists, nothing to do (and not checking its primary type for now): {}", fullPath);
-                } else {
+                final Node node;
+                if (session.nodeExists(fullPath)) {
+                    log.info("Node at {} already exists, checking/adjusting its types", fullPath);
+                    node = session.getNode(fullPath);
+                    if (psd.getPrimaryType() != null && !node.getPrimaryNodeType().getName().equals(psd.getPrimaryType())) {
+                        log.info("Adjusting primary type of node {} to {}", fullPath, psd.getPrimaryType());
+                        node.setPrimaryType(psd.getPrimaryType());
```

Comment:
IMHO this is a fundamental change to the way repo-init deals with the 'create path' statement and i am not entirely sure if it wouldn't be better to use a dedicated syntax for enforcing the primary type.
personally, i would prefer if the primary type would have been set/adjusted from the very first version of repoinit. but i fear that changing it now will cause regressions as consumers of repoinit may have flawed type information with create-path statements that just work because a given node already exists.... changing the behavior now would result in these mistakes to become visible and thus failing the repository start up.

Reply:
see https://github.com/apache/sling-org-apache-sling-jcr-repoinit/pull/41#issuecomment-1353043016